### DISCUSSION TITLE
Multiple schedulers with same RedisJobStore

### DISCUSSION THREAD

**Discussion by duy-nguyen-ts** [agronholm/apscheduler#915] [General]
**Hi everyone, I have this case and I want to know exactly how apscheduler handle it**
# Prepare
- apscheduler `version 3.10.4`
- Two schedulers A and B => describe multiple instance scheduler => using `AsyncIOScheduler()`
- Both using same RedisJobStore =>  do not change the job_key or anything else, just define and connect
- A function `test_job(name: str)`
# Do
- Initialize` A = AsyncIOScheduler()` and `B = AsyncIOScheduler()`
- `A.add_jobstore(RedisJobStore())` and `B.add_jobstore(RedisJobStore())`
- `A.start()` and `B.start()`
- Call `A.add_job(func=test_job, trigger='date', kwargs={"name": "AAA"}, run_date=t, id="same_id", replace_existing=True)`
- Call `B.add_job(func=test_job, trigger='date', kwargs={"name": "BBB"}, run_date=t+5, id="same_id", replace_existing=True)`

# Result
- What happen at `t+5` ? => I have tested and the job ran only one time with `name = BBB`
- I have added the listener and func `test_job` sometime run by scheduler A and sometime by scheduler B
- If I use MemoryStore instead of RedisJobStore => the job ran twice with different name => Expected behavior => It's ok
==> So I want to know if I add a job with same ID and same job_store between multiple instance. Is that job always run only one time with the latest `kwargs` ?

**Comment by agronholm** [comment]
Have you read the [FAQ](https://apscheduler.readthedocs.io/en/3.x/faq.html#how-do-i-share-a-single-job-store-among-one-or-more-worker-processes)? Job store sharing does not work on 3.x, but does on 4.0/master. Sadly it turns out Redis did not have sophisticated enough querying capabilities so it was dropped as a store in 4.0.

**Reply by duy-nguyen-ts** [reply]
Thank you @agronholm, I have read FAQ, but I think in above case the job will be duplicated and run twice. I don't know why it run only once so I need the confirmation 😄 
Btw, I have an app which has their own scheduler => I have two instances for it. Currently I using memory store, but I want to move to JobStore such as RedisJobStore to avoid missing job when restarting service. But when moving to RedisJobStore I will face with `JobStore sharing`. So how I can avoid it ?

**Reply by duy-nguyen-ts** [reply]
After testing again with starting 2 different processes, the job will run twice as expected. So how we can avoid `JobStore sharing` among multiple instances of schedulers ?

**Reply by italobbarros** [reply]
I have the exact same problem. My service has 2 instances, and I use Jobstore on PostgreSQL with SQLAlchemy, but the job runs twice. Did you find a solution to avoid this problem?

**Reply by hduynam99** [reply]
I ran into the same problem here, any solution yet ?
